linera-service/tests/fixtures/erc20_shared.sol

```diff
@@ -0,0 +1,317 @@
+// contract and so a modification of the code could not be avoided.
+//
+// This implementation has been adapted to our need in the following
+// ways:
+// * The totalSupply() of this contract is constant and fixed at
+//   construction to be the total one. It returns the sum of the supply
+//   over all the chains.
+// * The localSupply() has been introduced to return the supply of one
+//   particular chain.
+// * When tokens are transfered to another chain, they are first burned
+//   on the sender chain and then minted on the receiving chain.
+// * The token can be transferred to another chain but under the same
+//   user.
+
+pragma solidity ^0.8.20;
+
+import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+
+import "./Linera.sol";
+import "./LineraTypes.sol";
+
+/**
+ * @dev Implementation of the {IERC20} interface.
+ *
+ * This implementation is agnostic to the way tokens are created. This means
+ * that a supply mechanism has to be added in a derived contract using {_mint}.
+ *
+ * TIP: For a detailed writeup see our guide
+ * https://forum.openzeppelin.com/t/how-to-implement-erc20-supply-mechanisms/226[How
+ * to implement supply mechanisms].
+ *
+ * The default value of {decimals} is 18. To change this, you should override
+ * this function so it returns a different value.
+ *
+ * We have followed general OpenZeppelin Contracts guidelines: functions revert
+ * instead returning `false` on failure. This behavior is nonetheless
+ * conventional and does not conflict with the expectations of ERC-20
+ * applications.
+ */
+contract ERC20_shared is Context, IERC20, IERC20Errors {
+    mapping(address account => uint256) private _balances;
+
+    mapping(address account => mapping(address spender => uint256)) private _allowances;
+
+    uint256 private summed_total_supply; // total supply of all the chains.
+    uint256 private _totalSupply; // total supply of this chain
+
+    /**
+     * @dev Sets the values for {name} and {symbol}.
+     *
+     */
+    constructor(uint256 the_supply) {
+        summed_total_supply = the_supply;
+    }
+
+    /**
+     * This is the function run on the creator chain to create the initial supply.
+     */
+    function instantiate(bytes memory input) external {
+        uint256 input_value = abi.decode(input, (uint256));
+        require(input_value == summed_total_supply);
+        _mint(msg.sender, input_value);
+    }
+
+    /**
+     * This is the function run on the receiving chain in order to credit
+     * the user with the tokens.
+     */
+    function execute_message(bytes memory input) external {
+        (address user, uint256 amount) = abi.decode(input, (address, uint256));
+        _mint(user, amount);
+    }
+
+    /**
+     * This is the transfer to another chain under the same username.
+     */
+    function transferToChain(bytes32 chain_id, uint256 amount) external {
+        _burn(msg.sender, amount);
+        bytes memory message = abi.encode(msg.sender, amount);
```

Comment:
Why not support sending to a different chain but also a different account on the recipient chain? I understand we can simulate that by sending to the `msg.sender` on a different chain and then moving tokens to a new user within that chain but that's two transactions and the second one being on a chain where we might not be an owner even